src/main/java/org/opensearch/ad/ratelimit/CheckpointReadWorker.java

```diff
@@ -0,0 +1,383 @@
+            mediumSegmentPruneRatio,
+            lowSegmentPruneRatio,
+            maintenanceFreqConstant,
+            clientUtil,
+            CHECKPOINT_READ_QUEUE_CONCURRENCY,
+            executionTtl,
+            CHECKPOINT_READ_QUEUE_BATCH_SIZE,
+            stateTtl,
+            stateManager
+        );
+
+        this.modelManager = modelManager;
+        this.checkpointDao = checkpointDao;
+        this.entityColdStartQueue = entityColdStartQueue;
+        this.resultWriteQueue = resultWriteQueue;
+        this.indexUtil = indexUtil;
+        this.cacheProvider = cacheProvider;
+        this.checkpointWriteQueue = checkpointWriteQueue;
+        this.entityColdStarter = entityColdStarter;
+    }
+
+    @Override
+    protected void executeBatchRequest(MultiGetRequest request, ActionListener<MultiGetResponse> listener) {
+        clientUtil.<MultiGetRequest, MultiGetResponse>execute(MultiGetAction.INSTANCE, request, listener);
+    }
+
+    /**
+     * Convert the input list of EntityFeatureRequest to a multi-get request.
+     * RateLimitedRequestWorker.getRequests has already limited the number of
+     * requests in the input list. So toBatchRequest method can take the input
+     * and send the multi-get directly.
+     * @return The converted multi-get request
+     */
+    @Override
+    protected MultiGetRequest toBatchRequest(List<EntityFeatureRequest> toProcess) {
+        MultiGetRequest multiGetRequest = new MultiGetRequest();
+        for (EntityRequest request : toProcess) {
+            Optional<String> modelId = request.getModelId();
+            if (false == modelId.isPresent()) {
+                continue;
+            }
+            multiGetRequest.add(new MultiGetRequest.Item(CommonName.CHECKPOINT_INDEX_NAME, modelId.get()));
+        }
+        return multiGetRequest;
+    }
+
+    @Override
+    protected ActionListener<MultiGetResponse> getResponseListener(List<EntityFeatureRequest> toProcess, MultiGetRequest batchRequest) {
+        return ActionListener.wrap(response -> {
+            final MultiGetItemResponse[] itemResponses = response.getResponses();
+            Map<String, MultiGetItemResponse> successfulRequests = new HashMap<>();
+
+            // lazy init since we don't expect retryable requests to happen often
+            Set<String> retryableRequests = null;
+            Set<String> notFoundModels = null;
+            for (MultiGetItemResponse itemResponse : itemResponses) {
+                String modelId = itemResponse.getId();
+                if (itemResponse.isFailed()) {
+                    final Exception failure = itemResponse.getFailure().getFailure();
+                    if (failure instanceof IndexNotFoundException) {
+                        for (EntityRequest origRequest : toProcess) {
+                            // submit to cold start queue
+                            entityColdStartQueue.put(origRequest);
+                        }
+                        return;
```

Comment:
I am not quite sure, is it return? not continue? I saw there is for loop outside.

Reply:
It is return.  If it is checkpoint index not found exception, I don't need to retry as checkpoint read is bound to fail. Just send everything to the cold start queue and return. Added comments to explain.